README.md
# jCodeGen

  -An automatic Makefile generator, jAutoMakefile for C and C++ programs.
  -An automatic C++ class generator, jClassGen : creates a .hpp and .cpp files.


changelog.txt
19/01/18 : Creation of the jCodeGen directory
23/01/19 : -Upload of the source code of jAutoMakefile and jClassGen
	   -Creation of a better verision of jAutoMakefile, jAutoMk (in C++)


jAutoMk/auto_makefile.c
/*
** auto_makefile.c for my_automakefile in /home/jakaranga/delivery/AutoMakefile
** 
** Made by JakaRanga
** Login   <jeremy.bouton@epitech.eu>
** 
** Started on  Wed Mar 22 10:51:52 2017 JakaRanga
** Last update Mon Aug 21 15:12:57 2017 JakaRanga
*/

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <sys/types.h>
#include <dirent.h>

int	dot_h(char**);
int	my_strlen(char*);
int	my_strcmp(char*, char*);
char	*my_strcat(char*, char*);
char	*get_next_line(int);
char	**my_str_to_wordtab(char*, char);

void	put_error(int err)
{
  if (err == 1)
    printf("\033[31;40mError of malloc\033[00m\n");
  if (err == 2)
    printf("\033[31;40mError of open\033[00m\n");
}

char	*name_file(char *name)
{
  char	*res;
  int	i;
  int	a;

  i = -1;
  a = -1;
  while (name[++i] != '\0')
    if (name[i] == '/')
      a = i;
  if ((res = malloc(sizeof(char) * (my_strlen(name) - a + 1))) == NULL)
    return (put_error(2), NULL);
  i = 0;
  while (name[++a] != '\0')
    {
      res[i] = name[a];
      i++;
    }
  res[i] = '\0';
  return (res);
}

char	*add_libs(char *s, int *libs)
{
  if (libs[0] == 1 || libs[1] == 1 || libs[2] == 1)
    {
      if ((s = my_strcat(s, "LDFLAGS\t=\t-L./libs ")) == NULL)
	return (NULL);
      if (libs[0] == 1)
	{
	  if ((s = my_strcat(s, "-lptf ")) == NULL)
	    return (NULL);
	}
      if (libs[1] == 1)
	{
	  if ((s = my_strcat(s, "-lmy ")) == NULL)
	    return (NULL);
	}
      if (libs[2] == 1)
	{
	  if ((s = my_strcat(s, "-lgnl ")) == NULL)
	    return (NULL);
	}
      if ((s = my_strcat(s, "\n\n")) == NULL)
	return (NULL);
    }
  return (s);
}

int	auto_gen(char **av, int *libs, int mod)
{
  char	*s;
  int	fd;
  int	i;

  i = 1;
  open("Makefile", O_CREAT, S_IRWXU);
  if ((fd = open("Makefile", O_WRONLY)) == -1)
    return (put_error(2), 84);
  if ((s = malloc(sizeof(char) * 1)) == NULL)
    return (put_error(1), 84);
  s[0] = '\0';
  if ((s = my_strcat(s, "SRCS\t=\t")) == NULL)
    return (put_error(1), 84);
  while (av[++i] != NULL)
    {
      if ((s = my_strcat(s, av[i])) == NULL)
	return (put_error(1), 84);
      if (av[i + 1] != NULL)
	if ((s = my_strcat(s, "\t\t\\\n\t\t")) == NULL)
	  return (put_error(1), 84);
    }
  if (mod == 0)
    if ((s = my_strcat(s, "\n\nOBJS\t=\t$(SRCS:.c=.o)\n\n")) == NULL)
      return (put_error(1), 84);
  if (mod == 1)
    if ((s = my_strcat(s, "\n\nOBJS\t=\t$(SRCS:.cpp=.o)\n\n")) == NULL)
      return (put_error(1), 84);
  if ((s = my_strcat(s, "NAME\t=\t")) == NULL)
    return (put_error(1), 84);
  if ((s = my_strcat(s, av[1])) == NULL)
    return (put_error(1), 84);
  if (mod == 0)
    if ((s = my_strcat(s, "\n\nCC\t=\tgcc\n\n")) == NULL)
      return (put_error(1), 84);
  if (mod == 1)
    if ((s = my_strcat(s, "\n\nCC\t=\tg++\n\n")) == NULL)
      return (put_error(1), 84);    
  if (libs[3] == 1)
    if ((s = my_strcat(s, "CFLAGS\t=\t-Wall -Wextra -Werror\n\n")) == NULL)
      return (put_error(1), 84);
  if (mod == 0)
    if ((s = add_libs(s, libs)) == NULL)
      return (put_error(1), 84);
  if ((s = my_strcat(s, "$(NAME):\t$(OBJS)\n\t\t$(CC) $(OBJS) $(LDFLAGS) -o $(NAME)\n\n")) == NULL)
    return (put_error(1), 84);
  if ((s = my_strcat(s, "all:\t\t$(NAME)\n\n")) == NULL)
    return (put_error(1), 84);
  if ((s = my_strcat(s, "re:\t\tfclean all\n\n")) == NULL)
    return (put_error(1), 84);
  if ((s = my_strcat(s, "clean:\n\t\trm -f $(OBJS)\n\n")) == NULL)
    return (put_error(1), 84);
  if ((s = my_strcat(s, "fclean:\t\tclean\n\t\trm -f $(NAME)\n\n")) == NULL)
    return (put_error(1), 84);
  if ((s = my_strcat(s, ".PHONY:\t\tall clean fclean re\n")) == NULL)
    return (put_error(1), 84);
  i = my_strlen(s);
  write(fd, s, i);
  free(s);
  return (printf("\033[32;40mMakefile created, just put your header in and add your own libs!\nMade by Jeremy BOUTON (Jaka)\033[00m\n"), 0);
}

int	is_name(char *s)
{
  int	i;

  i = -1;
  while (s[++i] != '\0')
    if (s[i] == '.' && s[i + 1] == 'c')
      return (84);
  return (0);
}

char	*get_name(char **env)
{
  int	i;
  int	j;
  int	line;
  char	*name;

  i = -1;
  line = -1;
  while (env[++i] != NULL)
    {
      if (env[i][0] == 'U' && env[i][1] == 'S' && env[i][2] == 'E' && env[i][3] == 'R')
	line = i;
    }
  if (line == -1)
    return ("USER");
  if ((name = malloc(sizeof(char) * my_strlen(env[line]))) == NULL)
    return (put_error(1), "USER");
  i = 5;
  j = 0;
  while (env[line][i] != '\0')
    {
      name[j] = env[line][i];
      j++;
      i++;
    }
  name[j] = '\0';
  return (name);
}

int	main(int ac, char **av, char **env)
{
  int	mod;
  char	*s;
  int	*libs;
  char	*tmp;

  mod = 0;
  if ((tmp = get_name(env)) == NULL)
    return (84);
  printf("\033[35;40mHello %s!\033[00m\n", tmp);
  free(tmp);
  if ((s = name_file(av[0])) == NULL)
    return (put_error(2), 84);
  if (ac < 3 || is_name(av[1]) == 84)
    return (printf("\033[32;40mUsage: ./%s [binary name] [files]\nMade by Jeremy BOUTON (Jaka)\033[00m\n", s), free(s), 0);
  else
    printf("Would is Make a C or C++ program ?\n");
  if ((s = get_next_line(0)) == NULL)
    return (put_error(2), 84);
  if (my_strcmp(s, "C++") == 0 || my_strcmp(s, "c++") == 0 || my_strcmp(s, "cpp") == 0)
    mod = 1;
  if ((libs = malloc(sizeof(int) * 4)) == NULL)
    return (84);
  libs[0] = 0, libs[1] = 0, libs[2] = 0, libs[3] = 0;
  printf("\033[34;01mWould you add libs?  (y / n)\033[00m\n");
  if (my_strcmp(get_next_line(0), "n") != 0)
    {
      if (mod == 0)
	{
	  printf("libptf.a (my_printf) ?\n (y / n)\n");
	  if (my_strcmp(get_next_line(0), "n") == 0)
	    libs[0] = -1;
	  else
	    libs[0] = 1;
	  printf("libmy.a ?\n (y / n)\n");
	  if (my_strcmp(get_next_line(0), "n") == 0)
	    libs[1] = -1;
	  else
	    libs[1] = 1;      
	  printf("libgnl.a ?\n (y / n)\n");
	  if (my_strcmp(get_next_line(0), "n") == 0)
	    libs[2] = -1;
	  else
	    libs[2] = 1;
	}
    }
  printf("Flags -Wall -Wextra -Werror ?\n (y / n)\n");
  if (my_strcmp(get_next_line(0), "n") == 0)
    libs[3] = -1;
  else
    libs[3] = 1;
  if (auto_gen(av, libs, mod) == 84)
    return (84);
  free(libs);
  return (0);
}


jMakeGenerator/main.cpp
//
// EPITECH PROJECT, 2018
// make gen
// File description:
// 
//

#include "AutoMk.hpp"

void	usage(int mod)
{
	if (mod == 1)
	{
		printf("\033[22;32m");
		std::cout << "\tUsage: ./jAutoMakefile [binary name] [files]" << std::endl;
	}
	if (mod == 2)
	{
		printf("\033[01;31m");
		std::cout << "\t/!\\ERROR! Your binary name is not correct!" << std::endl;
	}
	if (mod == 3)
	{
		printf("\033[01;31m");
		std::cout << "\t/!\\ ERROR! A Makefile is already created!" << std::endl;
	}
	printf("\033[22;32m");
	std::cout << "Made by Jeremy BOUTON (JakaRanga)" << std::endl;
	printf("\033[01;37m");
}

int	main(int ac, char **av)
{
	AutoMk *mk;

	mk = new AutoMk(ac, av);
	mk->sayHello();
	if (!mk->verifyParams())
		return (usage(mk->getMod()), 84);
	if (!mk->writeFile())
		return (usage(mk->getMod()), 84);
	else
		usage(0);
	return 0;
}


jMakeGenerator/AutoMk.cpp
//
// EPITECH PROJECT, 2018
// AutoMk.cpp
// File description:
//
//

#include "AutoMk.hpp"

AutoMk::AutoMk(const int &ac, char **av)
{
	this->Ac = ac;
	this->Av = av;
	this->Mod = 0;
	this->isCpp = true;
}

int		AutoMk::getMod()
{
	return this->Mod;
}

std::string	AutoMk::getUserName()
{
	std::string	name;

	name = std::getenv("USER");
	if (name[0] >= 'a' && name[0] <= 'z')
		name[0] -= 32;
	return name;
}

void		AutoMk::sayHello()
{
	printf("\033[01;36m");
	std::cout << "Hello " << this->getUserName() << "!" << std::endl;
	printf("\033[01;37m");
}

bool		AutoMk::verifyParams()
{
	int		i = 0;
	
	if (this->Ac < 3)
		return ((this->Mod = 1), false);
	while (this->Av[1][i++])
		if (this->Av[1][i] == '.')
			return ((this->Mod = 2), false);
	for (i = 2; this->Av[i]; i++)
	{
		std::string	str(this->Av[i]);
		if (str.find(".cpp") == std::string::npos)
			this->isCpp = false;
		else
			this->isCpp = true;
	}
	return true;
}

bool		AutoMk::fileExist()
{
	std::ifstream	tmp("Makefile");

	if (tmp.is_open())
	{
		std::string     gLine;
		std::size_t     y;

		printf("\033[01;31m");
		std::cout << "Do you really want overwrite the existing Makefile?"
			  << std::endl;
		printf("\033[01;37m");
		std::cout << "\t(N / y)" << std::endl;

		std::getline(std::cin, gLine);
		y = gLine.find("y");

		if (y == 0)
			return (false);
		else
			return ((this->Mod = 3), true);
	}
	tmp.close();
	return false;
}

std::ofstream	&AutoMk::putHeader(std::ofstream &mk)
{
	mk << "##" << std::endl << "## EPITECH PROJECT, 2018" << std::endl
           << "## Makefile" << std::endl
           << "## File description:" << std::endl
           << "##" << std::endl << "##" << std::endl << std::endl;
							
	return mk;
}

bool		AutoMk::writeFile()
{
	if (this->fileExist())
		return false;
	
	std::ofstream	mk("Makefile");

	this->putHeader(mk);

	mk << "SRCS\t=\t" << this->Av[2] << "\t\\" << std::endl;
	for (int i = 3; this->Av[i]; i++)
	{
		mk << "\t\t" << this->Av[i]
		   << ((this->Av[i + 1] != NULL) ? "\t\\" : "") << std::endl;
	}
	mk << std::endl << "OBJS\t=\t$(SRCS:" << ((this->isCpp) ? ".cpp" : ".c") << "=.o)"
	   << std::endl << std::endl << "NAME\t=\t" << this->Av[1] << std::endl
	   << std::endl << "CC\t=\t" << ((this->isCpp) ? "g++" : "gcc") << std::endl
	   << std::endl << "LDFLAGS\t=\t-W -Wall -Wextra -Werror" << std::endl
	   << std::endl << "$(NAME):\t$(OBJS)" << std::endl << "\t\t"
	   << "$(CC) $(OBJS) $(LDFLAGS) -o $(NAME)" << std::endl
	   << std::endl << "all:\t\t$(NAME)" << std::endl
	   << std::endl << "re:\t\tfclean all" << std::endl
	   << std::endl << "clean:" << std::endl << "\t\trm -f $(OBJS)" << std::endl
	   << std::endl << "fclean:\t\tclean" << std::endl << "\t\trm -f $(NAME)"
	   << std::endl << std::endl << ".PHONY:\t\tall clean fclean re";

	mk.close();
	printf("\033[22;32m");
	std::cout << "\tYour Makefile have been generated!" << std::endl;
	printf("\033[01;37m");
	return true;
}

AutoMk::~AutoMk()
{
}


jClassGenerator/ClassGenerator.cpp
//
// JAKA PROJECT, 2018
//
// File description:
// Cpp file jClassGenerator
//

#include "ClassGenerator.hpp"

ClassGenerator::ClassGenerator(const std::string &name)
{
	this->Name = name;
	this->Head = getHpp();
	this->User = pickEnvName();
}

std::string		ClassGenerator::pickEnvName()
{
	std::string	name;

	name = std::getenv("USER");
	if (name[0] >= 'a' && name[0] <= 'z')
		name[0] -= 32;
	return name;
}

std::string		ClassGenerator::getUserName()
{
	return this->User;
}

std::string		ClassGenerator::getHpp()
{
	std::string	head = this->Name;
	int		i = 0;

	while (head[i])
	{
		if (head[i] >= 'a' && head[i] <= 'z')
			head[i] -= 32;
		i++;
	}
	head += "_HPP_";
	return (head);
}

bool			ClassGenerator::fileExist()
{
	std::ifstream	fHpp(this->Name + ".hpp");
	std::ifstream	fCpp(this->Name + ".cpp");

	if (fHpp.is_open() || fCpp.is_open())
	{
		std::string	gLine;
		std::size_t	y;

		printf("\033[01;31m");
		std::cout << "Do you really want overwrite the existing class ["
			  << this->Name << "]?" << std::endl;
		printf("\033[01;37m");
		std::cout << "\t(N / y)" << std::endl;

		std::getline(std::cin, gLine);
		y = gLine.find("y");

		if (y == 0)
			return true;
		else
			return false;
	}
	return true;
}

std::ofstream			&ClassGenerator::putHeader(std::ofstream &pp, bool isCpp)
{
	pp << "//" << std::endl << "// EPITECH PROJECT, 2018" << std::endl
	   << "// " << this->Name << ((isCpp) ? ".cpp" : ".hpp") << std::endl
	   << "// File description:" << std::endl
	   << "//" << std::endl << "//" << std::endl << std::endl;;
	return pp;
}

bool			ClassGenerator::fileGen()
{
	
	std::ofstream	fHpp(this->Name + ".hpp");
	std::ofstream	fCpp(this->Name + ".cpp");

	if (!fHpp || !fCpp)
		return false;

	putHeader(fHpp, false);
	putHeader(fCpp, true);
	
	fHpp << "#ifndef " << this->Head << std::endl
	     << "# define " << this->Head << std::endl << std::endl
	     << "#include <iostream>" << std::endl << std::endl
	     << "class " << this->Name << std::endl
	     << "{" << std::endl << "public:" << std::endl
	     << "\t" << this->Name << "();" << std::endl << std::endl
	     << "\t~" << this->Name << "();" << std::endl << std::endl
	     << "private:" << std::endl
	     << "};" << std::endl << std::endl
	     << "#endif /* !" << this->Head << " */" << std::endl;

	fCpp << "#include \"" << this->Name << ".hpp\"" << std::endl << std::endl
	     << this->Name << "::" << this->Name << "()" << std::endl
	     << "{" << std::endl << "}" << std::endl << std::endl << std::endl
	     << this->Name << "::~" << this->Name << "()" << std::endl
	     << "{" << std::endl << "}";

	fHpp.close();
	fCpp.close();
	return true;
}

ClassGenerator::~ClassGenerator()
{
}
